ts/src/index.ts

```diff
@@ -57,18 +61,21 @@ export class AutoPush {
 
   async preprocessRequest(
       reqPath: string, stream: http2.ServerHttp2Stream,
-      cacheCookie?: string): Promise<string> {
+      cacheCookie?: string): Promise<PreprocessResult> {
     const cacheChecker = cacheCookie ?
         ClientCacheChecker.deserialize(cacheCookie) :
         new ClientCacheChecker();
     // Calculate the auto-push list before sending the response to be able to
     // set the bloom filter cookie correctly that contains the auto-pushed
     // assets as well as the original asset. Otherwise we'll auto-push assets
     // that browser already has in future responses.
-    this.pushList = await this.getAutoPushList(reqPath, stream, cacheChecker);
+    const pushList = await this.getAutoPushList(reqPath, stream, cacheChecker);
     cacheChecker.addPath(reqPath);
-    const cacheCookieValue = cacheChecker.serialize();
-    return cacheCookieValue;
+    const newCacheCookie = cacheChecker.serialize();
+    return {
+      newCacheCookie,
+      pushFn: async (stream) => await this.push(stream, pushList),
```

Comment:
You can avoid creating an async function here and calling await. I would just do `push: (stream) =>  this.push( stream, pushList)`.
This would save you several promise allocations.

Reply:
Done. I would hope that v8 optimizes away the async/await pair but maybe not.